Mathlib/RingTheory/GradedAlgebra/FGHomogeneousIdeal.lean

```diff
@@ -0,0 +1,166 @@
+2. Prove that the span of the spanning set we have constructed is indeed the original
+   homogeneous ideal.
+-/
+
+variable {ι A σ : Type*}
+variable [Semiring A]
+variable [DecidableEq ι] [AddMonoid ι]
+variable [SetLike σ A] [AddSubmonoidClass σ A]
+variable (𝒜 : ι → σ) [GradedRing 𝒜] (I : HomogeneousIdeal 𝒜) (hI : Ideal.FG I.toIdeal)
+
+/--
+For each `a : A`, `GradedRing.homogeneousComponents 𝒜 a` is the collection of the
+homogeneous components of `a`, which is a finite subset of `A`.
+-/
+def GradedRing.homogeneousComponents [DecidableEq A] (a : A) : Finset A :=
+  Finset.image (λ i ↦ (DirectSum.decompose 𝒜 a) i) (DFinsupp.support (DirectSum.decompose 𝒜 a))
+
+lemma GradedRing.homogeneousComponents_def [DecidableEq A] (a : A) :
+    GradedRing.homogeneousComponents 𝒜 a = Finset.image (λ i ↦ (DirectSum.decompose 𝒜 a) i)
+    (DFinsupp.support (DirectSum.decompose 𝒜 a)) := rfl
+
+lemma GradedRing.ne_zero_of_mem_homogeneousComponents [DecidableEq A] (a c : A) :
+    c ∈ GradedRing.homogeneousComponents 𝒜 a → c ≠ 0 := λ hc ↦ by
+  rw [homogeneousComponents, Finset.mem_image] at hc; rcases hc with ⟨i, hi1, hi2⟩; subst hi2;
+  rw [DFinsupp.mem_support_iff] at hi1; simp only [ne_eq, ZeroMemClass.coe_eq_zero]; exact hi1
+
+lemma GradedRing.exists_of_mem_homogeneousComponents [DecidableEq A] (a c : A) :
+    c ∈ GradedRing.homogeneousComponents 𝒜 a → ∃ (i : ι), DirectSum.decompose 𝒜 a i = c :=
+  λ hc ↦ by
+  rw [GradedRing.homogeneousComponents] at hc;
+  simp only [Finset.mem_image, DFinsupp.mem_support_toFun, ne_eq] at *;
+  rcases hc with ⟨i, _, _⟩; use i
+
+lemma GradedRing.mem_homogeneousComponents_of_ne_zero_and_exists [DecidableEq A] (a c : A) :
+    (c ≠ 0 ∧ ∃ (i : ι), DirectSum.decompose 𝒜 a i = c) →
+    c ∈ GradedRing.homogeneousComponents 𝒜 a := λ ⟨hc1, i, hi⟩ ↦ by
+  rw [homogeneousComponents, Finset.mem_image]; exact ⟨i, by
+  rw [DFinsupp.mem_support_iff]; subst hi; exact Subtype.ne_of_val_ne hc1, hi⟩
+
+lemma GradedRing.mem_homogeneousComponents_iff [DecidableEq A] (a c : A) :
+    c ∈ GradedRing.homogeneousComponents 𝒜 a ↔
+    c ≠ 0 ∧ ∃ (i : ι), DirectSum.decompose 𝒜 a i = c :=
+  ⟨λ hc ↦ ⟨ne_zero_of_mem_homogeneousComponents 𝒜 a c hc, exists_of_mem_homogeneousComponents
+  𝒜 a c hc⟩, λ hc ↦ mem_homogeneousComponents_of_ne_zero_and_exists 𝒜 a c hc⟩
+
+lemma GradedRing.mem_homogeneousSubmonoid_of_mem_homogeneousComponents [DecidableEq A]
+    (a c : A) (hc : c ∈ GradedRing.homogeneousComponents 𝒜 a) :
+    c ∈ SetLike.homogeneousSubmonoid 𝒜 := by
+  rw [← (GradedRing.exists_of_mem_homogeneousComponents 𝒜 a c hc).choose_spec]
+  dsimp [SetLike.homogeneousSubmonoid]
+  simp only [Submonoid.mem_mk, Subsemigroup.mem_mk, Set.mem_setOf_eq, SetLike.homogeneous_coe]
+
+namespace HomogeneousIdeal
+
+/--
+A finite subset of `A` which spans the finitely generated ideal `I.toIdeal`.
+-/
+noncomputable def FG.spanningSet : Finset A := hI.choose
+
+lemma FG.spanningSet_span_eq : Ideal.span (FG.spanningSet 𝒜 I hI) = I.toIdeal := hI.choose_spec
```

Comment:
This has nothing to do with `HomogeneousIdeal`, maybe this should be moved to another file as well.